### PR TITLE
Auto distribute count based on Max brick size

### PR DESCRIPTION
**Pull request by aravindavk**
While creating auto provisioned volume, support added to automatically
calculate the distribute count based on max brick size specified in
the request.

For example, below command creates 2x3(Distributed replicate) volume

```
glustercli volume create gv1 --replica 3 --size 1G \
        --max-brick-size 512M
```

Fixes: #999
Signed-off-by: Aravinda VK <avishwan@redhat.com>

### CODE REVIEW

File: glustercli/cmd/volume-create.go [hunk 1]
```diff
@@ -102,10 +104,16 @@ func smartVolumeCreate(cmd *cobra.Command, args []string) {
 		failure("Invalid File Size specified", nil, 1)
 	}
 
+	maxBrickSize, err := sizeToBytes(flagCreateMaxBrickSize)
```
Comment (Madhu-1) [inline glustercli/cmd/volume-create.go]: if `flagCreateMaxBrickSize` has changed then do size calculation?

File: e2e/smartvol_ops_test.go [hunk 2]
```diff
@@ -407,6 +407,54 @@ func testSmartVolumeDistributeDisperse(t *testing.T) {
 	checkZeroLvs(r)
 }
 
+func testSmartVolumeAutoDistributeReplicate(t *testing.T) {
```
Comment (Madhu-1) [inline e2e/smartvol_ops_test.go]: can you add some negative test cases?

Comment (aravindavk) [reply e2e/smartvol_ops_test.go]: what kind of negative tests I can add? negative max-brick-size or max-brick-size is greater than request size?

File: glustercli/cmd/volume-create.go (hunk 1 above)
Comment (aravindavk) [reply glustercli/cmd/volume-create.go]: sizeToBytes returns zero when input is empty, no conversion/calculation done

File: e2e/smartvol_ops_test.go (hunk 2 above)
Comment (Madhu-1) [reply e2e/smartvol_ops_test.go]: kind of invalid minimum and maximum values for  max-brick-size

**Comment by aravindavk** [comment]
@Madhu-1 please review

**Comment by aravindavk** [comment]
Found one more issue in Volume create flow, opened issue https://github.com/gluster/glusterd2/issues/1389

File: pkg/lvmutils/utils.go [hunk 3]
```diff
@@ -345,3 +341,12 @@ func ExtendThinpool(expansionTpSizePerBrick uint64, vgName string, tpName string
 	err := utils.ExecuteCommandRun("lvextend", fmt.Sprintf("-L+%dB", expansionTpSizePerBrick), fmt.Sprintf("/dev/%s/%s", vgName, tpName))
 	return err
 }
+
+// NormalizeSize converts the value to multiples of 512
```
Comment (Madhu-1) [inline pkg/lvmutils/utils.go]: can we make it unexported function?

File: glusterd2/bricksplanner/planner.go [hunk 4]
```diff
@@ -95,6 +87,37 @@ func getBricksLayout(req *api.VolCreateReq) ([]api.SubvolReq, error) {
 		return nil, err
 	}
 
+	if req.MaxBrickSize > 0 && req.MaxBrickSize < minBrickSize {
+		return nil, errors.New("invalid max-brick-size, Minimum size required is " + strconv.Itoa(minBrickSize))
+	}
+
+	// If max Brick size is specified then decide distribute
+	// count and Volume Size based on Volume Type
+	if req.MaxBrickSize > 0 && req.Size > req.MaxBrickSize {
```
Comment (Madhu-1) [inline glusterd2/bricksplanner/planner.go]: what happens if `req.MaxBrickSize` if greater than `req.Size`?

Comment (aravindavk) [reply glusterd2/bricksplanner/planner.go]: It will be same as not passing that option

File: pkg/lvmutils/utils.go (hunk 3 above)
Comment (aravindavk) [reply pkg/lvmutils/utils.go]: this is used in `glusterd2/bricksplanner` as well

File: glusterd2/bricksplanner/planner.go (hunk 4 above)
Comment (Madhu-1) [reply glusterd2/bricksplanner/planner.go]: as this is an input request we need to return an error to the user.

Comment (aravindavk) [reply glusterd2/bricksplanner/planner.go]: I don't think that is an error. It is satisfying the user's condition of max-brick-size.

File: e2e/smartvol_ops_test.go [hunk 5]
```diff
@@ -407,6 +407,83 @@ func testSmartVolumeDistributeDisperse(t *testing.T) {
...
+	// Max-brick-size is more than request size
+	createReq = api.VolCreateReq{
+		Name:               smartvolname,
+		Size:               20 * gutils.MiB,
+		ReplicaCount:       3,
+		MaxBrickSize:       30 * gutils.MiB,
+		SubvolZonesOverlap: true,
+	}
+	volinfo, err = client.VolumeCreate(createReq)
+	r.Nil(err)
```
Comment (atinmu) [inline e2e/smartvol_ops_test.go]: Shouldn't this be r.NotNil (err) , how is this going to succeed if the MaxBrickSize is greater than Size ?

[the remark this answers is not in the document]
Comment (aravindavk) [reply e2e/smartvol_ops_test.go]: max subvol size is derived based on volume type and `req.MaxBrickSize`. This calculation is same for Arbiter/Replica. Special handling only for Disperse volume.

Comment (aravindavk) [reply e2e/smartvol_ops_test.go]: `MaxBrickSize > VolumeSize` condition will always satisfy. So this satisfies user's condition.

**Comment by aravindavk** [comment]
> How would this max brick size be leveraged in PVC request? I see that if this isn't passed then we don't decide on the distribute count.

Two ways we can distribute now,

- Specify distribute count in the Volume create request
- Specify max-brick-size and calculate the distribute count automatically.

Max brick size can be added in Storage class definition irrespective of Volume sizes. That will be passed with every volume create request. (For example, `max-brick-size=10GiB`). If Volume requested of size less than that value will be created without distribute.

Currently `max-brick-size` takes precedence over the `distribute-count` parameter in Volume create request. That means if request contains `distribute=3`, `max-brick-size=512M` and `size=1G` then it will create volume with distribute count 2(Distribute count specified in request is ignored). Let me know if you feel this precedence needs to be reversed.(If distribute count is specified then do not consider `max-brick-size`)

**Comment by atinmu** [comment]
@aravindavk I'm not sure if distributeCount and maxBricksSize parameters input value should co-exist but if they do then I agree that maxBrickSize should take precedence as distribute count should be more internal and should be managed intelligently by GD2.

**Comment by atinmu** [comment]
@aravindavk @Madhu-1 don't we need a respective change in the gluster-csi-driver once this gets in? If so, I think we need an issue tracking this.

**Comment by Madhu-1** [comment]
@atinmu currently we are just making use of size and volume type in volume create,we will make required changes in CSI driver.

**Comment by rishubhjain** [comment]
Fixes: #1398

**Comment by aravindavk** [comment]
retest this please

**Pull request merged**